feat(WASI-NN, piper): update the WASI-NN Piper plugin due to upstream changes

Divyansh200102 opened the pull request:
Fixes #4436

### Description
This PR updates the WASI-NN Piper plugin dependencies to reflect the upstream migration from the archived `rhasspy/piper` repository to the new active maintenance repository at `OHF-Voice/piper1-gpl`, addressing Issue #4436.

### Changes
* **Dependency Update:** Updated `cmake/WASINNDeps.cmake` to fetch from `OHF-Voice/piper1-gpl` instead of `rhasspy/piper`.
* **Version Bump:** Updated the git tag to `v1.3.0` (commit `fee9b9c`).
* **Test Artifacts:** Updated download URLs in `test/plugins/wasi_nn/CMakeLists.txt` for `test_voice.onnx`, `test_voice.onnx.json`, and the Piper binary to point to the new repository and release tags.

### Notes
- Do not merge before #4481

Divyansh200102 commented:
> Could you check if the C++ alternative, libpiper, is ready for use? If the upstream doesn't have such pre-built assets, we can also build it within the CI pipeline and statically link it into the plugin.

@hydai I think it's ready for use they have it in forked here along with the documentation on how to build espeak as well as  download shared libraries for onnxruntime . https://github.com/roboticslab-uc3m/piper1-gpl/blob/rl/libpiper/README.md
But I don't see it in the official repo. https://github.com/OHF-Voice/piper1-gpl/tree/v1.3.0
They do have installation guides as well. https://robots.uc3m.es/installation-guides/install-piper.html

>This is also expected, I believe we have to update the patch to fit the new upstream.

Should I update the patch in this PR only or create another one ?

Divyansh200102 commented:
I think it's ready for use. They also have instructions on how to build espeak-ng and download shared libraries for onnxruntime. https://github.com/OHF-Voice/piper1-gpl/tree/main/libpiper

hydai commented on .github/workflows/reusable-build-extensions-on-linux.yml:
```diff
@@ -67,8 +67,27 @@ jobs:
+            PIPER_INSTALL_PATH="$PWD/install"
+            echo "--- Verify install ---"
+            ls -R "$PIPER_INSTALL_PATH"
+            cd ../..
```
We should move these lines into a standalone script, maybe called `build_libpiper.sh`, and use it instead of exporting everything here.

hydai commented on .github/workflows/reusable-build-extensions-on-macos.yml:
```diff
@@ -61,7 +61,27 @@ jobs:
+            ls -R "$PIPER_INSTALL_PATH"
+            cd ../..
+          fi
+          # ---------------------------------------
```
These lines are just duplicated.

hydai commented on test/plugins/wasi_nn/CMakeLists.txt:
```diff
@@ -88,25 +88,31 @@ foreach(BACKEND ${WASMEDGE_PLUGIN_WASI_NN_BACKEND})
     message(STATUS "Download ML artifacts to ${CMAKE_CURRENT_BINARY_DIR}/wasinn_piper_fixtures")
     download(
-      https://github.com/rhasspy/piper/raw/master/etc/test_voice.onnx
+       https://github.com/OHF-Voice/piper1-gpl/raw/main/tests/test_voice.onnx
```
Why do we need this space?

Divyansh200102 replied:
I'll revert it back.

Divyansh200102 commented:
@hydai I am getting this error in the CI while building piper plugin https://github.com/Divyansh200102/WasmEdge/actions/runs/20550549997/job/59027578939#logs 
 
```
plugins/wasi_nn/CMakeFiles/wasmedgePluginWasiNN.dir/wasinnmodule.cpp.o -c /__w/WasmEdge/WasmEdge/plugins/wasi_nn/wasinnmodule.cpp
In file included from /__w/WasmEdge/WasmEdge/plugins/wasi_nn/wasinnmodule.cpp:4:
In file included from /__w/WasmEdge/WasmEdge/plugins/wasi_nn/wasinnmodule.h:6:
In file included from /__w/WasmEdge/WasmEdge/plugins/wasi_nn/wasinnenv.h:14:
/__w/WasmEdge/WasmEdge/plugins/wasi_nn/wasinn_piper.h:11:10: fatal error: 'piper.hpp' file not found
   11 | #include <piper.hpp>
      |          ^~~~~~~~~~~
1 error generated.
[115/141] Building CXX object lib/api/CMakeFiles/wasmedgeCAPI.dir/wasmedge.cpp.o
[116/141] Building CXX object lib/vm/CMakeFiles/wasmedgeVM.dir/vm.cpp.o
ninja: build stopped: subcommand failed.
Error: Process completed with exit code 1.
0s
```

Divyansh200102 commented:
@hydai, could you please review this PR again? I’ve updated the PR to match the recent merges and fix the dependency issues we talked about. I switched the CI to use the libpiper install path that matches the Dockerfile to keep things consistent. I also updated `cmake/WASINNDeps.cmake` so it looks for the local onnxruntime (the one Piper installs) before trying the system one. This should take care of those version compatibility problems and make sure everything links up right.

0yi0 commented:
Hi @Divyansh200102, you can consider making a PR to your own `master` branch to check CI result first.

Divyansh200102 commented:
Changes:
- Removed the old patch file as dependencies like `piper phenomize ` were not needed anymore as we build espeak-ng  directly in the build wheel and introduced a new one to build libpiper statically.

The build is still failing as libpiper doesn't have the `piper.hpp` file which was a part of previous upstream . the new upstream has a similar file `piper_impl.hpp` so now in order to build libpiper we would need to modfiy the source code according to the new implementation. (https://github.com/OHF-Voice/piper1-gpl/blob/main/libpiper/include/piper_impl.hpp) 

CI jobs: https://github.com/Divyansh200102/WasmEdge/actions/runs/20971724746/job/60276643631?pr=13
https://github.com/Divyansh200102/WasmEdge/actions/runs/20971724746/job/60276643549?pr=13
Related file : `plugins\wasi_nn\wasinn_piper.h`

Divyansh200102 commented:
Hi @hydai , PTAL 

## Reproduction Steps

### 1. Build WasmEdge from source

```bash
cmake -GNinja -Bbuild -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_USE_LLVM=OFF -DWASMEDGE_PLUGIN_WASI_NN_BACKEND=Piper
cmake --build build
```

### 2. Prepare the test directory

* Create a new folder named `piper_test`.
* Copy the `espeak-data` folder from the build directory into this new folder.

### 3. Clone the example repository

```text
https://github.com/second-state/WasmEdge-WASINN-examples/tree/master/wasmedge-piper
```

### 4. Download the model and config files

```bash
# Download model
curl -LO https://huggingface.co/rhasspy/piper-voices/resolve/main/en/en_US/lessac/medium/en_US-lessac-medium.onnx

# Download config
curl -LO https://huggingface.co/rhasspy/piper-voices/resolve/main/en/en_US/lessac/medium/en_US-lessac-medium.onnx.json
```

### 5. Modify `main.rs` to generate a `.wav` file

Edit
`WasmEdge-WASINN-examples/wasmedge-piper/src/main.rs`

> If there is no `output_type` option, the default generated audio file will be `.wav`.

```rust
let config = serde_json::json!({
    "model": "en_US-lessac-medium.onnx", // path to .onnx voice file, required
    "config": "en_US-lessac-medium.onnx.json", // path to JSON voice config file, optional (default: model path + .json)
    "espeak_data": "./espeak-ng-data", // path to espeak-ng data directory, required for espeak phonemes
    "output_type": "wav"
});
```

### 6. Build and generate the `.wav` file

```bash
cargo build --target wasm32-wasip1 --release
cd ../..
```

```bash
LD_LIBRARY_PATH=$(pwd)/WasmEdge/build/lib \
WASMEDGE_PLUGIN_PATH=$(pwd)/WasmEdge/build/plugins/wasi_nn \
WasmEdge/build/tools/wasmedge/wasmedge \
--dir .:. \
WasmEdge-WASINN-examples/wasmedge-piper/target/wasm32-wasip1/release/wasmedge-piper.wasm
```

### Screenshot (WAV output)

<img width="1473" height="522" alt="WAV output screenshot" src="https://github.com/user-attachments/assets/8baa8ca3-55c0-4c9b-b5e5-f91bc0314bc0" />

---

## Generating a raw audio file

To generate a raw file, change `output_type` to `"raw"`.

### Screenshot (RAW output)

> Since this is a raw file, it cannot be played directly.

<img width="1477" height="525" alt="RAW output screenshot" src="https://github.com/user-attachments/assets/790a858c-7a08-4ecf-bfa0-99725ee25932" />

---

### CI jobs:
  [https://github.com/Divyansh200102/WasmEdge/actions/runs/21182102051/job/60927874828?pr=16](https://github.com/Divyansh200102/WasmEdge/actions/runs/21182102051/job/60927874828?pr=16)

Currently, the CI builds Piper from source instead of using the Docker prebuilt image. This is because `libpiper` was not built statically in the install script. Once the Docker image is updated, we can test the new implementation using the statically built `libpiper` dependency on Linux systems.

- Related PR:  #4535

As a result, we temporarily removed:

```text
-DPIPER_ROOT="$PIPER_INSTALL_PATH"
```

from:

```text
.github/workflows/reusable-build-extensions-on-linux.yml
```
- Once the docker image is updated and we can reintroduce this flag to check how prebuilt libpiper dependency used from the docker works with the current CI.
- Once this new implementation gets approved I will modfiy https://github.com/second-state/WasmEdge-WASINN-examples/blob/master/wasmedge-piper/README.md with the updated instructions to build piper from the new upstream.

hydai commented on plugins/wasi_nn/wasinn_piper.cpp:
```diff
@@ -72,19 +114,17 @@ WASINN::ErrNo parseSynthesisConfig(SynthesisConfig &SynthesisConfig,
-    if (auto Err =
-            getOptionalOption(Object, "speaker_id", SynthesisConfig.SpeakerId);
+  {
+    auto s_id = std::optional<int64_t>{};
```
Wrong coding style

hydai commented on plugins/wasi_nn/wasinn_piper.cpp:
```diff
@@ -487,42 +384,107 @@ Expect<WASINN::ErrNo> compute(WASINN::WasiNNEnvironment &Env,
+  } else {
+    size_t totalSize = audioBuffer.size() * sizeof(int16_t);
+    CxtRef.Output->resize(totalSize);
+    std::memcpy(CxtRef.Output->data(), audioBuffer.data(), totalSize);
```
We no longer accept `memcpy`. Please use `std::copy_n` instead.

hydai commented on plugins/wasi_nn/wasinn_piper.cpp:
```diff
@@ -7,10 +7,13 @@
 #include "simdjson.h"
 
+#include <algorithm>
+#include <cmath>
```
What is the reason we need these headers?

Divyansh200102 replied:
I've kept the algorithm header for std::copy_n and removed the unused map and cmath headers.

Divyansh200102 commented:
Apologies an `endif` was added by clang-tidy for some reason in wasinn_piper.h and because of that the build was failing I have fixed it. I have started the CI on my local fork when it passes I will link the result here and after that we can trigger the workflow here.

hydai commented on plugins/wasi_nn/wasinn_piper.cpp:
```diff
@@ -487,42 +384,109 @@ Expect<WASINN::ErrNo> compute(WASINN::WasiNNEnvironment &Env,
+      if (Sample > 1.0f)
+        Sample = 1.0f;
+      else if (Sample < -1.0f)
+        Sample = -1.0f;
```
Use `std::clamp` instead.
Ref: https://en.cppreference.com/w/cpp/algorithm/clamp.html

Divyansh200102 commented on plugins/wasi_nn/wasinn_piper.cpp:
```diff
@@ -487,42 +384,109 @@ Expect<WASINN::ErrNo> compute(WASINN::WasiNNEnvironment &Env,
+  std::vector<int16_t> AudioBuffer;
+  piper_audio_chunk Chunk;
+  int SampleRate = 0;
+  const float MaxWavValue = 32767.0f;
```
I agree with applying constexpr to the magic numbers like MaxWavValue, as it's the only value we know beforehand.
But all other variables depend on runtime data (like SampleRate or vector sizes ) so wouldn't changing them cause compilation error?

Divyansh200102 replied:
CI jobs: https://github.com/Divyansh200102/WasmEdge/actions/runs/21202443682?pr=16

hydai commented on plugins/wasi_nn/wasinn_piper.cpp:
```diff
@@ -487,42 +384,108 @@ Expect<WASINN::ErrNo> compute(WASINN::WasiNNEnvironment &Env,
+
+  while (piper_synthesize_next(GraphRef.Synth.get(), &Chunk) != PIPER_DONE) {
+    if (Chunk.num_samples == 0)
+      continue;
```
All 'if' blocks should be wrapped with `{}`.
```suggestion
    if (Chunk.num_samples == 0) {
      continue;
    }
```

Divyansh200102 commented:
Hi @hydai the CI checks for piper are passing now .

https://github.com/Divyansh200102/WasmEdge/actions/runs/21862246301/job/63098948043?pr=27
https://github.com/Divyansh200102/WasmEdge/actions/runs/21862246301/job/63098947853?pr=27


Reproduction steps

1. for using prebuilt piper and running the tests
```
./utils/wasi-nn/install-libpiper.sh

cmake -Bbuild -GNinja \
  -DCMAKE_BUILD_TYPE=Release \
  -DWASMEDGE_USE_LLVM=OFF \
  -DWASMEDGE_BUILD_TESTS=ON \
  -DWASMEDGE_BUILD_PLUGINS=ON \
  -DWASMEDGE_PLUGIN_WASI_NN_BACKEND="piper" \
  -DPIPER_ROOT="/usr/local" \
  .

cmake --build build --target wasmedgePluginWasiNN wasiNNTests

export ESPEAK_DATA_PATH="/usr/local/share/espeak-ng-data"

cd build

LD_LIBRARY_PATH=$(pwd)/lib/api ctest
``` 

2. Build from source

```
cmake -GNinja -Bbuild -DCMAKE_BUILD_TYPE=Release -DWASMEDGE_USE_LLVM=OFF -DWASMEDGE_PLUGIN_WASI_NN_BACKEND=Piper
cmake --build build
``` 

3. WASINN Example Run
Similar to https://github.com/WasmEdge/WasmEdge/pull/4443#issuecomment-3774513460

Please review this PR and see if it looks good merge.

hydai reviewed (approved):
LGTM.  
Thanks
